File: abzensi_online/lib/service/attendance_service/attendance_service.dart
import 'package:dio/dio.dart';
import 'package:hyper_ui/env.dart';
import 'package:hyper_ui/service/device_service/device_service.dart';
import 'package:hyper_ui/service/location_service/location_service.dart';

class AttendanceService {
  Future<(String, bool)> doSomething() async {
    return ("OK", false);
  }

  Future<(bool, String)> checkin({
    required String photo,
  }) async {
    try {
      var userDevice = await DeviceService.getDeviceModel();
      var locationResponse = await LocationService.getCurrentLocation();
      var position = locationResponse.position!;

      var response = await Dio().post(
        "${Env.baseUrl}/api/attendance/check-in",
        options: Env.options,
        data: {
          "check_in_device_id": userDevice,
          "check_in_latitude": position.latitude,
          "check_in_longitude": position.longitude,
          "check_in_photo": photo,
        },
      );
      Map data = response.data["data"];
      double distance = data["distance"];
      bool isRecognized = data["is_recognized"];
      bool isTooFar = data["is_too_far"];
      print("Distance: $distance");

      if (isTooFar) {
        return (false, "Jarak terlalu jauh");
      }

      if (!isRecognized) {
        return (false, "Wajah tidak dikenali");
      }

      return (true, "");
    } on Exception catch (_) {
      return (false, "Ada masalah pada API");
    }
  }

  Future<(bool, String)> checkOut({
    required String photo,
  }) async {
    try {
      var userDevice = await DeviceService.getDeviceModel();
      var locationResponse = await LocationService.getCurrentLocation();
      var position = locationResponse.position!;

      var response = await Dio().post(
        "${Env.baseUrl}/api/attendance/check-out",
        options: Env.options,
        data: {
          "check_out_device_id": userDevice,
          "check_out_latitude": position.latitude,
          "check_out_longitude": position.longitude,
          "check_out_photo": photo,
        },
      );
      Map data = response.data["data"];
      double distance = data["distance"];
      bool isRecognized = data["is_recognized"];
      bool isTooFar = data["is_too_far"];
      print("Distance: $distance");

      if (isTooFar) {
        return (false, "Jarak terlalu jauh");
      }

      if (!isRecognized) {
        return (false, "Wajah tidak dikenali");
      }

      return (true, "");
    } on Exception catch (_) {
      print(_);
      return (false, "Ada masalah pada API");
    }
  }

  Future<bool> isCheckInToday() async {
    try {
      var response = await Dio().post(
        "${Env.baseUrl}/api/attendance/is-check-in-today",
        options: Env.options,
      );
      Map obj = response.data;
      return obj["data"]["is_check_in_today"];
    } on Exception catch (_) {
      return false;
    }
  }

  Future<bool> isCheckOutToday() async {
    try {
      var response = await Dio().post(
        "${Env.baseUrl}/api/attendance/is-check-out-today",
        options: Env.options,
      );
      Map obj = response.data;
      return obj["data"]["is_check_out_today"];
    } on Exception catch (_) {
      return false;
    }
  }

  Future<List> getHistories() async {
    try {
      var response = await Dio().get(
        "${Env.baseUrl}/api/attendance/histories",
        options: Env.options,
      );
      Map obj = response.data;
      return obj["data"];
    } on Exception catch (_) {
      throw (_);
    }
  }

  Future resetToday() async {
    try {
      await Dio().post(
        "${Env.baseUrl}/api/attendance/reset-today",
        options: Env.options,
      );
    } on Exception catch (_) {}
  }
}

File: abzensi_online_api/app/Services/AttendanceService.php
<?php

namespace App\Services;

use App\Models\Attendance;

class AttendanceService
{
    public function checkIn($data)
    {
        $attendance = Attendance::create([
            'user_id' => $data['user_id'],
            'check_in_device_id' => $data['check_in_device_id'],
            'check_in_latitude' => $data['check_in_latitude'],
            'check_in_longitude' => $data['check_in_longitude'],
            'check_in_photo' => $data['check_in_photo'],
            'check_in_date' => now(),
        ]);
        return $attendance;
    }

    public function checkOut($data)
    {
        $attendance = Attendance::where("user_id", $data['user_id'])
            ->update([
                'check_out_device_id' => $data['check_out_device_id'],
                'check_out_latitude' => $data['check_out_latitude'],
                'check_out_longitude' => $data['check_out_longitude'],
                'check_out_photo' => $data['check_out_photo'],
                'check_out_date' => now(),
            ]);
        return $attendance;
    }

    public function isCheckInToday($user_id)
    {
        $todayAttendances = Attendance::where("user_id", $user_id)
            ->whereDate('check_in_date', now()->toDateString())->get();
        $isCheckIn = count($todayAttendances) > 0;
        return $isCheckIn;
    }

    public function isCheckOutToday($user_id)
    {
        $todayAttendances = Attendance::where("user_id", $user_id)
            ->whereNotNull('check_in_date')
            ->whereNotNull('check_out_date')
            ->get();
        $isCheckIn = count($todayAttendances) > 0;
        return $isCheckIn;
    }

    public function getHistories($user_id)
    {
        $todayAttendances = Attendance::where("user_id", $user_id)->get();
        return $todayAttendances;
    }
}
